x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/profile/ProfileServiceTests.java

```diff
@@ -306,6 +320,58 @@ public void testBuildSearchRequest() {
         }
     }
 
+    public void testSecurityProfileOrigin() {
+        // Activate profile
+        doAnswer(invocation -> {
+            assertThat(threadPool.getThreadContext().getTransient(ACTION_ORIGIN_TRANSIENT_NAME), equalTo(SECURITY_PROFILE_ORIGIN));
+            @SuppressWarnings("unchecked")
+            final ActionListener<SearchResponse> listener = (ActionListener<SearchResponse>) invocation.getArguments()[2];
+            listener.onResponse(SearchResponse.empty(() -> 1L, SearchResponse.Clusters.EMPTY));
+            return null;
+        }).when(client).execute(eq(SearchAction.INSTANCE), any(SearchRequest.class), anyActionListener());
+
+        when(client.prepareIndex(SECURITY_PROFILE_ALIAS)).thenReturn(
+            new IndexRequestBuilder(client, IndexAction.INSTANCE, SECURITY_PROFILE_ALIAS)
+        );
+
+        final RuntimeException expectedException = new RuntimeException("expected");
+        doAnswer(invocation -> {
+            assertThat(threadPool.getThreadContext().getTransient(ACTION_ORIGIN_TRANSIENT_NAME), equalTo(SECURITY_PROFILE_ORIGIN));
+            final ActionListener<?> listener = (ActionListener<?>) invocation.getArguments()[2];
+            listener.onFailure(expectedException);
+            return null;
+        }).when(client).execute(eq(BulkAction.INSTANCE), any(BulkRequest.class), anyActionListener());
+
+        final PlainActionFuture<Profile> future1 = new PlainActionFuture<>();
+        profileService.activateProfile(AuthenticationTestHelper.builder().realm().build(), future1);
+        final RuntimeException e1 = expectThrows(RuntimeException.class, future1::actionGet);
+        assertThat(e1, is(expectedException));
+
+        // Update
+        doAnswer(invocation -> {
+            assertThat(threadPool.getThreadContext().getTransient(ACTION_ORIGIN_TRANSIENT_NAME), equalTo(SECURITY_PROFILE_ORIGIN));
+            final ActionListener<?> listener = (ActionListener<?>) invocation.getArguments()[2];
+            listener.onFailure(expectedException);
+            return null;
+        }).when(client).execute(eq(UpdateAction.INSTANCE), any(UpdateRequest.class), anyActionListener());
+        final PlainActionFuture<UpdateResponse> future2 = new PlainActionFuture<>();
+        profileService.doUpdate(mock(UpdateRequest.class), future2);
+        final RuntimeException e2 = expectThrows(RuntimeException.class, future2::actionGet);
```

Comment:
What we actually want to test is the action origin is switched to `security_profile` for each profile related actions. This is done by the assertion in the `doAnswer`. What the lambda actually returns is irrelevant after the assertion is passed. Mocking a full successful response is possible but not useful here since they are no different than an exception with what the test tries to achieve. So I chose to keep it simple. But to address your point, I add a comment to the test method to explain what it does.